### PR TITLE
WIP Journal Project View #GSoC2016

### PR DESCRIPTION
**Pull request by samdroid-apps**
@AbrahmAB Journal Project View GSoC project

Required branches:

* sugar - this one
* sugar-artwork - AbrahmAB's's project-icon branch
* sugar-datastore - AbrahmAB's gsoc branch
IMHO, improvements that could be made are:

* [x] Remove black icon in the "add new" entry (what is that icon?)
* [x] Don't show in the project jobject in the project view's list (as per discussed previously)
* [x] "Choose an activity" popup - autofocus the search entry
* [x] "Choose an activity" popup - make search entry full popup width (entry.props.hexpand = True??)
* [x] "Choose an activity" popup - pressing enter should launch top activity
* [x] the icon to switch to the project view should be a radiotoolbutton (so that when you are in the project view, it has a lighter background it indicate it is toggled)

Probably a few more things, but it is looking great!

### CODE REVIEW

**Comment by AbrahmAB** [comment]
@samdroid-apps Changes made.

Re: Don't show in the project jobject in the project view's list (as per discussed previously)
use this sugar-datastore branch
https://github.com/AbrahmAB/sugar-datastore/tree/gsoc

ISSUE:
This works fine for project with non-zero objects(try adding activities to project)!

File: src/jarabe/desktop/activitychooser.py [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+        if self.tree_view.row_activated_handler:
+            self.tree_view.disconnect(self.tree_view.row_activated_handler)
+        if self.tree_view.button_press_handler:
+            self.tree_view.disconnect(self.tree_view.button_press_handler)
+        if self.tree_view.button_reslease_handler:
+            self.tree_view.disconnect(self.tree_view.button_reslease_handler)
+        if self.tree_view.icon_clicked_handler:
+            self.tree_view.disconnect(self.tree_view.icon_clicked_handler)
+
+        self.tree_view.connect('row-activated',self.__row_activated_cb)
```
Comment (samdroid-apps) [inline src/jarabe/desktop/activitychooser.py]: Hum, the button release handlers were added because the row-activated handler was only introduced in a newer Gtk+ version than available on OLPC OS.  It would be better maybe to override a "start-activity" signal from the activitylistview.

See the discussion on this pull request:  https://github.com/sugarlabs/sugar/pull/691

File: src/jarabe/desktop/activitychooser.py [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+        if len(toolbar.search_entry._icon_selected) == 1 \
+           and not toolbar.search_entry.get_text() == activity['name']:
+            pos = toolbar.search_entry.get_position()
+            toolbar.search_entry.set_text(
+                toolbar.search_entry._icon_selected[0]['name'])
+            toolbar.search_entry.select_region(pos, -1)
+
+    def __key_press_event_cb(self, widget, event):
+        if not self.search_bar.search_entry.has_focus():
+            self.search_bar.search_entry.grab_focus()
```
Comment (samdroid-apps) [inline src/jarabe/desktop/activitychooser.py]: Can you grab focus when the widget is shown?  (either in __init__ or under the realize signal, whichever works)

Right now, there is no indication that this behaviour happens - the user doesn't know that trying == grabbing focus.

File: src/jarabe/desktop/activitychooser.py [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+        self._query = normalize_string(query.decode('utf-8'))
+        self._list_view.set_filter(self._query)
+
+        toolbar.search_entry._icon_selected = \
+            self._list_view.get_activities_selected()
+
+        # verify if one off the selected names is a perfect match
+        # this is needed by th case of activities with names contained
+        # in other activities like 'Paint' and 'MusicPainter'
+        for activity in self._list_view.get_activities_selected():
```
Comment (samdroid-apps) [inline src/jarabe/desktop/activitychooser.py]: I feel like this code is copped from the home view.  That's good from a UX consistency standpoint.  But can you re-use the code without copy and paste?

File: src/jarabe/desktop/activitychooser.py [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+            toolbar.search_entry.set_text(
+                toolbar.search_entry._icon_selected[0]['name'])
+            toolbar.search_entry.select_region(pos, -1)
+
+    def __key_press_event_cb(self, widget, event):
+        if not self.search_bar.search_entry.has_focus():
+            self.search_bar.search_entry.grab_focus()
+
+        if widget == self.search_bar.search_entry:
+            if event.keyval == Gdk.KEY_Return and len(widget._icon_selected) == 1:
```
Comment (samdroid-apps) [inline src/jarabe/desktop/activitychooser.py]: `IconEntry object has no attribute _icon_selected` -- I get the error when pressing enter.

File: src/jarabe/desktop/activitychooser.py [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+    def __activitylist_clear_clicked_cb(self, list_view, toolbar):
+        toolbar.clear_query()
+
+    def set_title(self, text):
+        self.get_title_box().set_title(text)
+
+    def __row_activated_cb(self, treeview, path, col):
+        if col is not treeview.get_column(0):
+            model = treeview.get_model()
+        row = model[path]
```
Comment (samdroid-apps) [inline src/jarabe/desktop/activitychooser.py]: model is not defined?

File: src/jarabe/journal/journalactivity.py [hunk 6]
```diff
@@ -241,12 +255,45 @@ def _session_manager_shutdown_cb(self, event):
...
+        self._entry.add_clear_button()
+        self._add_new_box.pack_start(self._entry, True, True, 0)
+        self._entry.show()
+        return self._add_new_box
+
+    def get_add_new_box(self):
+        return self._add_new_box
+
+    def _create_add_new_entry_project(self):
+        hbox = Gtk.Box(orientation = Gtk.Orientation.HORIZONTAL)
```
Comment (samdroid-apps) [inline src/jarabe/journal/journalactivity.py]: the code is looking good from a macro level, just a few flake8 issues.  running flake8 prints them all out.

File: src/jarabe/journal/journalactivity.py [hunk 7]
```diff
@@ -241,12 +255,45 @@ def _session_manager_shutdown_cb(self, event):
...
         return False
 
-    def _setup_main_view(self):
-        self._main_toolbox = MainToolbox()
-        self._edit_toolbox = EditToolbox(self)
-        self._main_view = Gtk.VBox()
-        self._main_view.set_can_focus(True)
-
+    def _create_add_new_entry(self):
+        self._add_new_box = Gtk.Box(orientation = Gtk.Orientation.HORIZONTAL)
```
Comment (samdroid-apps) [inline src/jarabe/journal/journalactivity.py]: Can you make the add new box into a class?  I see this function copied around a lot.

File: src/jarabe/journal/journalactivity.py [hunk 8]
```diff
@@ -257,20 +304,91 @@ def _setup_main_view(self):
...
-        self._main_view.pack_start(self._list_view, True, True, 0)
-        self._list_view.show()
+        self._list_view.connect('project-view-activate',
+                                self.project_view_activated_cb)
+        self._list_view.tree_view.connect('choose-project',
+                                self.__choose_project_cb)
+        return self._list_view
+
 
+    def _create_list_view_project(self):
```
Comment (samdroid-apps) [inline src/jarabe/journal/journalactivity.py]: what does this function do differently from _create_list_view?

File: src/jarabe/journal/journalactivity.py [hunk 9]
```diff
@@ -257,20 +304,91 @@ def _setup_main_view(self):
...
+        logging.debug('project_view_activate signal handler')
+        self.set_canvas(self._project_view)
+        self._toolbox = self._main_toolbox
+        self.set_toolbar_box(self._toolbox)
+        self._toolbox.show()
+
+        query = {}
+        query['project_id'] = self.project_metadata['uid']
+        #query['mountpoints'] = ['/']
+        self._list_view_project.update_with_query(query)
```
Comment (samdroid-apps) [inline src/jarabe/journal/journalactivity.py]: Can we make the _list_view_project a responsability of the ProjectView?  So we give it to the ProjectView constructor, and the ProjectView deals with adding it to it's box, updating the query, etc.

File: src/jarabe/journal/journalactivity.py [hunk 10]
```diff
@@ -282,17 +400,70 @@ def _setup_secondary_view(self):
         self._secondary_view.pack_end(self._detail_view, True, True, 0)
         self._detail_view.show()
 
+    def _key_press_add_new_entry_cb(self, window, event):
```
Comment (samdroid-apps) [inline src/jarabe/journal/journalactivity.py]: sugar convention is double underscores for callbacks

File: src/jarabe/journal/journalactivity.py [hunk 11]
```diff
@@ -282,17 +400,70 @@ def _setup_secondary_view(self):
...
+        keyname = Gdk.keyval_name(event.keyval)
+        if keyname == 'Return':
+            self._add_new_button_clicked_cb(None, None)
+
+        return False
+
+    def _add_new_button_clicked_cb(self, button, event):
+        if self.get_list_view().get_projects_view_active():
+            logging.debug('[GSoC]for project title %r' %self.get_entry().props.text)
```
Comment (samdroid-apps) [inline src/jarabe/journal/journalactivity.py]: let's remove the [GSoC] loggine prefixs

File: src/jarabe/journal/journalactivity.py [hunk 12]
```diff
@@ -282,17 +400,70 @@ def _setup_secondary_view(self):
...
+                                      project_metadata=None)
+        elif self.project_metadata is not None:
+            chooser = ActivityChooser()
+            text = _("Choose an activity to start '%s' with" %self._entry_project.props.text)
+            chooser.set_title(text)
+            chooser.connect('activity-selected',self.__activity_selected_cb)
+            chooser.show_all()
+
+    def __activity_selected_cb(self, widget, bundle_id, activity_id):
+        initialize_journal_object(title = _(self._entry_project.props.text),
```
Comment (samdroid-apps) [inline src/jarabe/journal/journalactivity.py]: Don't pass the user's title through gettext

File: src/jarabe/journal/journalactivity.py [hunk 13]
```diff
@@ -282,17 +400,70 @@ def _setup_secondary_view(self):
...
+                                        bundle_id=bundle_id, 
+                                        activity_id=activity_id,
+                                        project_metadata=self.project_metadata)
+
     def _key_press_event_cb(self, widget, event):
-        if not self._main_toolbox.search_entry.has_focus():
-            self._main_toolbox.search_entry.grab_focus()
+        #if not self._main_toolbox.search_entry.has_focus():
+        #self._main_toolbox.search_entry.grab_focus()
```
Comment (samdroid-apps) [inline src/jarabe/journal/journalactivity.py]: can we restore the old behaviour of default-to-search-entry-focus.  So, the could would be:

    if the seacrh_entry doesn't have focus AND no other entry has focus:
        focus the search entry

File: src/jarabe/journal/journalactivity.py [hunk 14]
```diff
@@ -488,5 +666,59 @@ def get_journal():
...
+        activity_id = activityfactory.create_activity_id()
+
+    if bundle_id == PROJECT_BUNDLE_ID:
+        logging.debug('[GSoC]initializing journal object for project %r' %[title, invited])
+        
+        jobject = datastore.create()
+        jobject.metadata['title'] = title
+        jobject.metadata['title_set_by_user'] = '0'
+        jobject.metadata['activity'] = PROJECT_BUNDLE_ID
+        jobject.metadata['activity_id'] = activity_id
```
Comment (samdroid-apps) [inline src/jarabe/journal/journalactivity.py]: Can you share the shared lines between the PROJECT and non project if statements?

File: src/jarabe/journal/journalactivity.py [hunk 15]
```diff
@@ -282,17 +400,70 @@ def _setup_secondary_view(self):
...
+        #if not self._main_toolbox.search_entry.has_focus():
+        #self._main_toolbox.search_entry.grab_focus()
 
         keyname = Gdk.keyval_name(event.keyval)
         if keyname == 'Escape':
             self._main_toolbox.clear_query()
             self.show_main_view()
 
+    def __choose_project_cb(self, tree_view, metadata_to_copy):
+        project_chooser = ObjectChooser()
```
Comment (samdroid-apps) [inline src/jarabe/journal/journalactivity.py]: please set the filter to projects

File: src/jarabe/journal/journalactivity.py [hunk 16]
```diff
@@ -282,17 +400,70 @@ def _setup_secondary_view(self):
...
+
+    def __project_chooser_response_cb(self, project_chooser, reponse_value,
+                                    metadata_to_copy):
+        object_id = project_chooser.get_selected_object_id()
+        metadata = model.get(object_id)
+        title = metadata_to_copy.get('title', None)
+        activity = metadata_to_copy.get('activity', None)
+        activity_id = metadata_to_copy.get('activity_id', None)
+        logging.debug('metadata_to_copy %r'%metadata_to_copy)
+        initialize_journal_object(title=title, bundle_id=activity,
```
Comment (samdroid-apps) [inline src/jarabe/journal/journalactivity.py]: Are you actually copying the file content?  I can't get that to work :(

Maybe it would be better to have a move to project feature instead - same as what you have now, but just change the metadata of the existing jobject.

File: src/jarabe/journal/journalactivity.py [hunk 17]
```diff
@@ -479,6 +651,12 @@ def freeze_ui(self):
     def unfreeze_ui(self):
         self._set_widgets_sensitive_state(True)
 
+    def initialize_journal_object(self, title=None, bundle_id=None,
```
Comment (samdroid-apps) [inline src/jarabe/journal/journalactivity.py]: what is the point of this method?  Can't people just call initialize_journal_object?

File: src/jarabe/journal/listview.py [hunk 18]
```diff
@@ -415,6 +423,13 @@ def __cell_select_toggled_cb(self, cell, path):
 
     def update_with_query(self, query_dict):
         logging.debug('ListView.update_with_query')
+            
+        if 'activity' in query_dict and query_dict['activity'] == PROJECT_BUNDLE_ID:
+            logging.debug('[GSoC] projectsview set in listview')
+            self.set_projects_view_active(True)
```
Comment (samdroid-apps) [inline src/jarabe/journal/listview.py]: if statment is not needed - just pass bool to the function

File: src/jarabe/journal/listview.py [hunk 19]
```diff
@@ -679,6 +699,24 @@ def __detail_clicked_cb(self, palette, uid):
...
         self.emit('volume-error', message, severity)
 
+    def set_projects_view_active(self, projects_view_active):
+        self._projects_view_active = projects_view_active
+        text = 'Add new entry'
+        logging.debug('set_projects_view_active %r'%projects_view_active)
+        if self._journalactivity:
+            if self._projects_view_active:
+                logging.debug('set_projects_view_active')
+                text = 'Add new project'
```
Comment (samdroid-apps) [inline src/jarabe/journal/listview.py]: pass that string via gettext

File: src/jarabe/journal/listview.py [hunk 20]
```diff
@@ -679,6 +699,24 @@ def __detail_clicked_cb(self, palette, uid):
...
+                self._journalactivity.get_entry().set_placeholder_text(text)
```
Comment (samdroid-apps) [inline src/jarabe/journal/listview.py]: get what entry?  that method could have a better name

File: src/jarabe/journal/misc.py [hunk 21]
```diff
@@ -87,6 +88,12 @@ def get_icon_name(metadata):
         bundle_id = metadata.get('bundle_id', '')
 
     if bundle_id:
+        if bundle_id == PROJECT_BUNDLE_ID:
+            file_name = \
+            '/home/broot/sugar-build/build' + \
+            '/out/install/share/icons/sugar/scalable/mimetypes/project-box.svg'
```
Comment (samdroid-apps) [inline src/jarabe/journal/misc.py]: ah, that's why the icon is not showing for me.

Don't hardcode the filename.

File: src/jarabe/journal/model.py [hunk 22]
```diff
@@ -598,10 +600,12 @@ def get(object_id):
     """Returns the metadata for an object
     """
     if os.path.exists(object_id):
+        logging.debug('get os.path')
         stat = os.stat(object_id)
         metadata = _get_file_metadata(object_id, stat)
         metadata['mountpoint'] = _get_mount_point(object_id)
     else:
+        logging.debug('get no os.path')
```
Comment (samdroid-apps) [inline src/jarabe/journal/model.py]: Not sure what value these added logging statments provide

File: src/jarabe/journal/palettes.py [hunk 23]
```diff
@@ -111,6 +121,12 @@ def __init__(self, journalactivity, metadata, detail=False):
         copy_menu.connect('volume-error', self.__volume_error_cb)
         menu_item.set_submenu(copy_menu)
 
+        if not metadata.get('activity',None) == PROJECT_BUNDLE_ID:
+            menu_item = MenuItem(_('Copy to project..'), 'project-box')
```
Comment (samdroid-apps) [inline src/jarabe/journal/palettes.py]: 3dots please

File: src/jarabe/journal/projectview.py [hunk 24]
```diff
@@ -0,0 +1,150 @@
...
+from jarabe.journal import model
+
+from sugar3.graphics import style
+from sugar3.graphics.toolbutton import ToolButton
+
+_SERVICE_NAME = 'org.laptop.Activity'
+_SERVICE_PATH = '/org/laptop/Activity'
+_SERVICE_INTERFACE = 'org.laptop.Activity'
+
+class ProjectView(Gtk.VBox):
```
Comment (samdroid-apps) [inline src/jarabe/journal/projectview.py]: Can you reuse some code from the DetailsView here?

I would love for the "top" of the project view (with the title and description box) to look *exactly* the same as the "top" of the details view.  Consistency is great.

So the top of the project view should have the same title entry box like this:

screenshot from 2016-07-09 14-42-02


and the same description textbox/label style.

File: src/jarabe/view/service.py [hunk 25]
```diff
@@ -71,7 +71,6 @@ def ActivateActivity(self, activity_id):
         activity already open.
         """
         activity = self._shell_model.get_activity_by_id(activity_id)
-
```
Comment (samdroid-apps) [inline src/jarabe/view/service.py]: please undo this change

File: src/jarabe/model/invites.py [hunk 26]
```diff
@@ -297,4 +297,4 @@ def get_instance():
     global _instance
     if not _instance:
         _instance = Invites()
-    return _instance
+    return _instance
```
Comment (samdroid-apps) [inline src/jarabe/model/invites.py]: please undo this change

File: src/jarabe/journal/journalactivity.py (hunk 15 above)
Comment (AbrahmAB) [inline src/jarabe/journal/journalactivity.py]: I have set filter in line below
`project_chooser._toolbar._proj_list_button_clicked_cb(None)`

File: src/jarabe/journal/journalactivity.py (hunk 16 above)
Comment (AbrahmAB) [inline src/jarabe/journal/journalactivity.py]: So are you suggesting me to implement 'Send to Project' instead of 'Copy to project'?

File: src/jarabe/journal/projectview.py (hunk 24 above)
Comment (AbrahmAB) [inline src/jarabe/journal/projectview.py]: Sounds like a good plan :)
Working on it.

**Comment by samdroid-apps** [comment]
Hi @AbrahmAB,

I think we can merge this when we fix 3 things:

* [ ] Project icon needs to be a little bigger
* [ ] Re-use the journal details views
* [ ] Use "show_message" on the list view when there is no entries in a project - hiding the issue in the list view

Also, a little bug:

* [ ] If you search for an activity in the ActivityChooser and press enter, it launches the activity that is at the top of the list when it is unfiltered.  Eg. 1) I have "abacus" installed, 2) I search "write", 3) I can see write at the top of the list, 4) I press enter, 5) Abacus activity launches

Also please run flake8 at some point.

I'd love to get this merged this arvo if we could!

File: src/jarabe/desktop/activitychooser.py (hunk 4 above)
Comment (AbrahmAB) [inline src/jarabe/desktop/activitychooser.py]: the bug of activity launching when activitychooser has more than one activity can be resolved by adding condition `len(widget._icon_selected) == 1`

Comment (AbrahmAB) [inline src/jarabe/desktop/activitychooser.py]: sorry my bad!! :( I am also getting same error! working on it

**Comment by samdroid-apps** [comment]
I love the re-use of the details view, it looks slick with the cleanup you did!

Thanks for fixing that bug in the activity chooser.  Is it also possible for pressing "enter" in the search bar to force a search (eg. not having to wait for the autosearch timeout?)

**Comment by samdroid-apps** [comment]
Also, another little bug, the list view doesn't refresh for me :(

Steps to reproduce:

**Comment by AbrahmAB** [comment]
Fixed :)

**Comment by samdroid-apps** [comment]
Hi Abhijit, just 2 more little things:

> * [ ]  Use "show_message" on the list view when there is no entries in a project - hiding the issue in the list view
> * [ ]  Project icon needs to be a little bigger (be *centred* in the space)

**Comment by AbrahmAB** [comment]
@samdroid-apps Changes made.
But whenever I try to re-size the icon and again run sugariconify.py it becomes B&W!!

**Comment by samdroid-apps** [comment]
@AbrahmAB, what I usually do to edit an icon is:

* open it in inkscape
* change the stroke to red and the fill to blue - this removes the entities
* edit it
* make it B&W
* sugar-iconify

The support in inkscape for the svg entities isn't the best it seems.

**Comment by samdroid-apps** [comment]
Hi @AbrahmAB.

I'm going to merge this, I just found 1 more bug:

* Clicking the "add to project..." menu entry doesn't show the object chooser for me.  Nothing in the log either :cry:.  Can you reproduce that or is it something in my setup?

**Comment by AbrahmAB** [comment]
@samdroid-apps  Bug fixed :)
- Improved the add-new entry code by implementing a common class in journaltoolbox.py as AddNewBar
